keras/engine/saving.py

```diff
@@ -95,106 +75,88 @@ def get_json_type(obj):
-            if isinstance(model.optimizer, optimizers.TFOptimizer):
-                warnings.warn(
-                    'TensorFlow optimizers do not '
-                    'make it possible to access '
-                    'optimizer attributes or optimizer state '
-                    'after instantiation. '
-                    'As a result, we cannot save the optimizer '
-                    'as part of the model save file.'
-                    'You will have to compile your model again '
-                    'after loading it. '
-                    'Prefer using a Keras optimizer instead '
-                    '(see keras.io/optimizers).')
+    model_weights_group = f['model_weights']
+    model_layers = model.layers
+    model_weights_group['layer_names'] = [layer.name.encode('utf8') for layer in model_layers]
+    model_weights_group['backend'] = K.backend().encode('utf8')
+    model_weights_group['keras_version'] = str(keras_version).encode('utf8')
+    for layer in model_layers:
+        layer_group = model_weights_group[layer.name]
+        symbolic_weights = layer.weights
+        weight_values = K.batch_get_value(symbolic_weights)
+        weight_names = []
+        for i, (w, val) in enumerate(zip(symbolic_weights, weight_values)):
+            if hasattr(w, 'name') and w.name:
+                name = str(w.name)
             else:
-                f.attrs['training_config'] = json.dumps({
-                    'optimizer_config': {
-                        'class_name': model.optimizer.__class__.__name__,
-                        'config': model.optimizer.get_config()
-                    },
-                    'loss': model.loss,
-                    'metrics': model.metrics,
-                    'sample_weight_mode': model.sample_weight_mode,
-                    'loss_weights': model.loss_weights,
-                }, default=get_json_type).encode('utf8')
-
-                # Save optimizer weights.
-                symbolic_weights = getattr(model.optimizer, 'weights')
-                if symbolic_weights:
-                    optimizer_weights_group = f.create_group(
-                        'optimizer_weights')
-                    weight_values = K.batch_get_value(symbolic_weights)
-                    weight_names = []
-                    for i, (w, val) in enumerate(zip(symbolic_weights,
-                                                     weight_values)):
-                        # Default values of symbolic_weights is /variable
-                        # for Theano and CNTK
-                        if K.backend() == 'theano' or K.backend() == 'cntk':
-                            if hasattr(w, 'name'):
-                                if w.name.split('/')[-1] == 'variable':
-                                    name = str(w.name) + '_' + str(i)
-                                else:
-                                    name = str(w.name)
+                name = 'param_' + str(i)
```

Comment:
Since we don't make constraint on naming, is it possible to have a duplicate name here? (Not sure if weight name need to be unique)

Reply:
I do share the same concern.